File: kraken/src/main/java/glitch/kraken/services/StreamService.java
package glitch.kraken.services;

import com.google.gson.JsonObject;
import glitch.api.http.HttpRequest;
import glitch.api.http.Routes;
import glitch.auth.objects.json.Credential;
import glitch.kraken.GlitchKraken;
import glitch.kraken.object.enums.StreamType;
import glitch.kraken.object.json.Game;
import glitch.kraken.object.json.Stream;
import glitch.kraken.object.json.StreamSummary;
import glitch.kraken.services.request.FeatureStreamsRequest;
import glitch.kraken.services.request.FollowedStreamRequest;
import glitch.kraken.services.request.LiveStreamsRequest;
import glitch.service.AbstractHttpService;
import reactor.core.publisher.Mono;


public class StreamService extends AbstractHttpService {
    public StreamService(GlitchKraken rest) {
        super(rest.getClient(), rest.getHttpClient());
    }

    public Mono<Stream> getStreamByUser(Long id, StreamType streamType) {
        HttpRequest request = Routes.get("/streams/%s").newRequest(id);

        if (streamType != null) {
            request.queryParam("stream_type", streamType.name().toLowerCase());
        }

        return exchangeTo(request, JsonObject.class).map(object -> object.get("stream"))
                .flatMap(element -> {
                    if (element.isJsonNull()) {
                        return Mono.empty();
                    } else {
                        return Mono.just(http.getGson().fromJson(element, Stream.class));
                    }
                });
    }

    public LiveStreamsRequest getLiveStreams() {
        return new LiveStreamsRequest(http);
    }

    public Mono<StreamSummary> getStreamSummary(Game game) {
        HttpRequest request = Routes.get("/streams/summary").newRequest();

        if (game != null) {
            request.queryParam("game", game.getName());
        }

        return exchangeTo(request, StreamSummary.class);
    }

    public FeatureStreamsRequest getFeaturedStreams() {
        return new FeatureStreamsRequest(http);
    }

    public FollowedStreamRequest getFollowedStreams(Credential credential) {
        return new FollowedStreamRequest(http, credential);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/StreamService.kt
package io.glitchlib.v5.service

import io.glitchlib.GlitchClient
import io.glitchlib.auth.Credential
import io.glitchlib.auth.Scope
import io.glitchlib.internal.http.body
import io.glitchlib.internal.http.bodyFlowable
import io.glitchlib.internal.http.bodySingle
import io.glitchlib.model.OrdinalList
import io.glitchlib.v5.internal.AbstractKrakenService
import io.glitchlib.v5.model.StreamType
import io.glitchlib.v5.model.json.FeatureStream
import io.glitchlib.v5.model.json.Game
import io.glitchlib.v5.model.json.Stream
import io.glitchlib.v5.model.json.StreamSummary
import io.glitchlib.v5.service.request.FeatureStreamsRequest
import io.glitchlib.v5.service.request.FollowedStreamRequest
import io.glitchlib.v5.service.request.LiveStreamsRequest


class StreamService(client: GlitchClient) : AbstractKrakenService(client) {

    fun getLiveStreams(request: LiveStreamsRequest.() -> Unit = {}) =
        get<OrdinalList<Stream>>("/streams", LiveStreamsRequest().apply(request)()).bodyFlowable

    fun getFreaturedStreams(request: FeatureStreamsRequest.() -> Unit = {}) =
        get<OrdinalList<FeatureStream>>("/streams/featured", FeatureStreamsRequest().apply(request)()).bodyFlowable

    fun getStreamByUser(id: Long, streamType: StreamType? = null) =
        get<OrdinalList<Stream>>("/streams/$id") {
            if (streamType != null) {
                addQueryParameters("stream_type", streamType.name.toLowerCase())
            }
        }.body.map { it.data[0] }

    fun getStreamSummary(game: Game? = null) =
        get<StreamSummary>("/streams/summary") {
            if (game != null) {
                addQueryParameters("game", game.name)
            }
        }.bodySingle

    fun getFollowedStreams(credential: Credential, request: FollowedStreamRequest.() -> Unit = {}) =
        if (credential.scopeCheck(Scope.USER_READ))
            get<OrdinalList<Stream>>(
                "/streams/followed",
                FollowedStreamRequest(credential).apply(request)()
            ).bodyFlowable
        else scopeIsMissing<Stream>(Scope.USER_READ).toFlowable()
}
